\documentclass{article}
\usepackage[round]{natbib}

\usepackage{fullpage}
\usepackage{listings}
\usepackage{url}

\usepackage{authblk}
\usepackage{graphicx}
\usepackage{color}

\lstset{language=Python}

% local definitions
\newcommand{\msprime}[0]{\texttt{msprime}}
\newcommand{\ms}[0]{\texttt{ms}}
\newcommand{\stdpopsim}[0]{\texttt{stdpopsim}}
\newcommand{\tskit}[0]{\texttt{tskit}}

\newcommand{\aprcomment}[1]{{\textcolor{blue}{APR: #1}}}
\newcommand{\dncomment}[1]{{\textcolor{red}{Dom: #1}}}
\newcommand{\sgcomment}[1]{{\textcolor{red}{SG: #1}}}
\newcommand{\jkcomment}[1]{{\textcolor{magenta}{JK: #1}}}

\begin{document}

\title{Describing demographic models is hard}
\author[ ]{A permutation of}
\author[1]{Aaron P. Ragsdale}
\author[1]{Dominic Nelson}
\author[2]{Jerome Kelleher}
\author[1]{Simon Gravel}
\affil[1]{McGill University and Genome Qu\'{e}bec Innovation Centre,
McGill University, Montr\'{e}al, Qu\'{e}bec, Canada}
\affil[2]{Big Data Institute, Li Ka Shing Centre for Health Information and Discovery,
University of Oxford, Oxford, United Kingdom}
\maketitle


\abstract{
Simulation plays a central role in population genomics studies. Recent years
have seen rapid improvements in software efficiency that make it possible to simulate
large genomic regions for many individuals sampled from large numbers of populations.
The increase in complexity of possible demographic models also provides additional ways
that we can get their implementation wrong. Here we describe two errors made in
defining population genetic models using the \msprime\ coalescent simulator that have
found their way into the published record.
We discuss how these errors have affected analyses and suggest recommendations
for software developers and users to reduce the risk of such errors.
}

\section{Introduction}

The \msprime\ coalescent simulator~\citep{kelleher2016efficient,kelleher2020coalescent}
is now quite widely used. The large increases in efficiency over the classical
\ms\ program~\citep{hudson2002generating} make it feasible to simulate large
samples of whole chromosomes for the first time. Another distinct advantage
of \msprime\ is the Python API that is its primary interface, greatly
increasing the flexibility and ease of use over the standard approach of
text-based command line interfaces. In particular, programs like \ms\
required users to specify cryptic command line options to describe demographic
models \jkcomment{maybe include an example ms command line?}. Particularly for
the large models we use today, these are not intuitive or comprehensible for humans.
The Python API for \msprime\ is a great improvement, allowing the user to
state models in a documented and reproducible manner. \jkcomment{maybe show
the same model as described above in msprime notation?}

Implementing multi-population models of demographic history is still hard and error
prone, however. In this note we discuss two implementation errors that arose through
unfortunate design decisions in \msprime's demography API and which then found their way
into the scientific record. While one of these errors is not expected to have produced qualitative
mistakes, the second lead to more severe ramifications \aprcomment{departures?}. In light of these
implementation errors, we discuss improvements to \msprime's API motivated by these
discoveries and best practices for implementing and simulating complex multi-population
demography.

\section{A bad tutorial example}

To illustrate the demography API, \msprime\ included a description of a widely-used
three population Out-of-Africa model~\citep{gutenkunst2009inferring}
as part of its tutorial documentation. In this model (Fig.~\ref{fig:ooa_stats}A),
Eurasian (CEU and CHB) and African (YRI) populations split from each other in the deep past,
followed by a more recent split of European and Asian populations, with variable rates of
continuous migration between each of the populations. However, the implementation in the
\msprime\ tutorial was incorrect. Namely, before the time of the split of African and Eurasian
populations, when there should have been just a single randomly mating population, migration was
allowed to occur between the ancestral population and a second population with size equal to
the Eurasian bottleneck size for all time into the past (Fig.~\ref{fig:ooa_stats}B).

%migration between those two populations was allowed to continue for all time into the
%past. Even though the population split was implemented correctly, the specified demographic model had 
%an additional population prior to the earliest split exchanging migrants with the ancestral population 
%(Fig.~\ref{fig:ooa_stats}B) but not contributing otherwise to present-day populations.  

Fortunately, the effects of this error are subtle. Population sizes and structure since the time of
the split are unaffected, so that differences in expected $F_{ST}$ are negligible between
the correct and incorrect model. However, the ancient structure distorts the distribution
of $T_{MRCA}$ beyond the split date and increases
heterozygosity in contemporary populations roughly $4\%$ over expectations
from the correct model (Fig.~\ref{fig:ooa_stats}E-F).

%However, the presence of the ghost population increases
%the effective population size \aprcomment{not true - the effect is a bit different than just change in Ne}
%and thus affects the distribution of older $T_{MRCA}$ and the expected 
%heterozygosity in each of the three populations by up to $4\%$ (Fig.~\ref{fig:ooa_stats}C-G).

This model was subsequently copied
several times and used in publications \jkcomment{do we want to estimate how
many times it was copied? A quick search on github suggests around 30
times]} \dncomment{I like this as a rough metric, like ``the number of copies
is unknown, but we found x copies on github''}

Such errors can be prevented by improved API design choice.
To model a population split currently in \msprime, a user must specify a
\texttt{MassMigration} event that moves lineages from one population to another
and then must also remember to turn off migration
between those populations at the same time.
The release of \msprime~ver.~1.0 will introduce a \texttt{PopulationSplit} event,
which more intuitively links the merger of lineages with appropriate changes in
migration rates at the time of the split.

\section{A missing parameter}

In another publication using this model~\citep{martin2017human},
a separate error was introduced: the model itself was defined as suggested
in the documentation (using updated parameters
from~\citet{gravel2011demographic}) and inspected
using the \msprime\ debugging tools.
After these initial checks were made, however, the simulation
was performed without passing the parameter of demographic events,
so that the three populations remained separated with low levels of migration and
never merged as expected (Fig~\ref{fig:prs}A),
leading to a vast overestimate of the divergence across human populations. 
Whereas the correct model predicts a mean $F_{ST}$ of
$0.05 - 0.10$ across the three population, the simulated model generated $F_{ST}$
ranging between $0.3 - 0.6$, depending on the populations considered.
Overall diversity was also strongly affected: expected heterozygosity was more than
doubled in African populations but just $30-40\%$ the expected
levels in Eurasian populations when compared to expectations from the correct model.

This simulation was performed to assess the
transferability of polygenic risk scores across human populations, in particular to
explore how human demographic history and population structure affects our ability
to predict genetic risk in populations of different ancestry than the originally studied
population.
The excess divergence exaggerated the role that demography plays
in limiting the transferability of genetic risk scores across populations.
The resulting publication has been influential in the discussion of
health inequalities and genomics, with over 350 citations since 2017.
We note that while difficulties in transferability remain in the corrected model
(Fig.~\ref{fig:prs}), risk prediction in each population is significantly improved
when using the correct demographic model
(compare to Fig.~5 in~\citet{martin2017human}).

These errors suggests three lessons.
First, the design of user interface and API for scientific software matters,
and many bugs can be prevented by using more intuitive interfaces.
Whereas the original \msprime\ required the user to pass
three separate parameters to specify a demographic model, \msprime~ver.~1.0
will introduce a Demography class, which wraps these three parameters.
Thus, instead of writing
\begin{lstlisting}[frame=single]
dbg = msprime.DemographyDebugger(
  population_configurations=population_configurations,
  migration_matrix=migration_matrix,
  demographic_events=demographic_events)
dbg.print_history()
ts = msprime.simulate(
  population_configurations=population_configurations,
  migration_matrix=migration_matrix,
  demographic_events=demographic_events)
\end{lstlisting}
with the error-inducing need to re-enter parameter information,  
we would now write
\begin{lstlisting}[frame=single]
demography = msprime.Demography(
  population_configurations=population_configurations,
  migration_matrix=migration_matrix,
  demographic_events=demographic_events)
dbg = demography.debug()
dbg.print_history()
ts = msprime.simulate(demography=demography)
\end{lstlisting}

Second, testing of the final simulated data is important.
This can be challenging, because the amount of simulated data can be large.
However, recent progress in computing summary statistics from tree
sequence data can make this easier~\citep{ralph2020efficiently}.
Even if the entire simulation data cannot be easily evaluated, subsets of the data
can be examined to identify large errors.

Third, open data analysis pipelines are necessary for the self-correcting nature of science.
This subtle error of large effect was only discovered through our own re-use of
the simulation pipeline developed in~\citet{martin2017human} to pursue
additional analyses on a similar topic. We identified the bug by tracing down an unexpected
result in our analysis. This error could not have been found and corrected without the open
publication of the entire analysis pipeline from the original study.

Finally, from a genetics perspective, the corrected simulations indicate that
the accuracy of genetic risk scores is still substantially reduced in understudied
populations (Fig.~\ref{fig:prs}E-G), supporting the main conclusion of~\citet{martin2017human}.
However, the reduction is much less pronounced than reported.
In particular, we do not observe large differences in mean risk prediction
across populations (Fig.~\ref{fig:prs}C,D), as presented in~\citet{martin2017human}.
At least for the neutral polygenic architectures considered here, genetic drift alone
does not appear to induce large biases in mean predicted risk across ethnicities.

\section{Conclusions}

The implementation of complex demographic models is error prone, and such errors
can have a large impact on downstream analysis and interpretation.
Recognition of these difficulties drove efforts to peer-review demographic model
implementation in \stdpopsim~\citep{adrion2019community}, with quality-controlled
models and simulation resources for a growing number of commonly studied species.

The first bug discussed in this note was indeed identified through a peer-review effort of the 
model implementation, and it is not easily identified by inspecting the resulting data.
The second bug would have been comparably easy to identify by inspection
of the data, but the missing parameter was easy to overlook. 

We therefore recommend the following steps to ensure more robust simulations.
First, we recommend that demographic models used in publications 
are verified by a second pair of eyes in code review or an independent implementation.
If the implementation is complex enough that code review is impractical, the risk
that bugs are introduced is extremely high and additional caution and statistical
verification is warranted.
Second, we recommend basic statistical verification of all simulated data.
If simulations are so large that statistical validation is burdensome,
subsets of the simulated data can be analyzed~\citep{ralph2020efficiently}.
Such tests would have caught the error in~\citet{martin2017human}.

Third, openness is essential. We only know about these errors because of open code and
open source development processes. By making their entire pipeline available,
\citet{martin2017human} not only enabled other research teams to build upon their findings,
but they made it possible for such errors to be found and corrected.
There must be many, many more mistakes out there, and we need both
pre- and post-publication vigilance from users and developers to ensure the
soundness of the large body of simulation-based analyses. 

\section{Methods}


We computed the expected allele frequency spectrum using
\texttt{moments}~ver.~1.0.3~\citep{jouganous2017inferring}, and LD-decay curves using
\texttt{moments.LD}~\citep{ragsdale2019models}. $F_{ST}$ and other diversity statistics
were computed from the expected AFS and verified using branch statistics from the
tree sequence records of \msprime\ simulations~\citep{ralph2020efficiently}.
Demographic models were plotted using the \texttt{demography} package
(\url{https://github.com/apragsdale/demography}, ver.~0.0.3).
We used the original pipeline from~\citet{martin2017human} available from
\url{https://github.com/armartin/ancestry_pipeline/blob/master/simulate_prs.py}.
We updated the pipeline to run with the correct demographic parameters and more
recent versions of \msprime\ and \tskit\, and the updated pipeline is
available at \url{https://github.com/apragsdale/PRS}.
Data and python scripts to recreate Figures~1~and~2 can be found at
\url{https://github.com/jeromekelleher/msprime-model-errors}.

\section*{Acknowledgements}

\begin{itemize}
\item Alicia Martin
\item 
\end{itemize}

\bibliographystyle{plainnat}
\bibliography{paper}

\pagebreak

\begin{figure}[ht]
\begin{center}
\makebox[\textwidth][c]{\includegraphics{figures/ooa_expected_stats.pdf}}
\caption{\textbf{Expected diversity statistics under the \citet{gutenkunst2009inferring} model}.
    \textbf{(A)} The correctly implemented model.
    \textbf{(B)} The incorrectly implemented model from the \msprime\ tutorial, with migration continuing
    into the past beyond the mass migration event with proportion $1$ from the ancestral population
    to the bottleneck population.
    \textbf{(C)} Marginal allele frequency spectra under the two models. Heterozygosity in the incorrect model
    is inflated by $~3.5\%$, though the general shape of the distributions are qualitatively similar.
    \textbf{(D)} Similarly, the increased heterozygosity leads to excess $D^2$, though the LD-decay is
    qualitatively similar between models.
    \textbf{(E-G)} True size history for each population plotted against the expected size history from
    the expected inverse coalescence rates.
}
\label{fig:ooa_stats}
\end{center}
\end{figure}

\begin{figure}[ht]
\begin{center}
\makebox[\textwidth][c]{\includegraphics{figures/prs_fig.pdf}}
\caption{\textbf{The transferability of PRS under neutrality}.
    \textbf{(A)} In~\citet{martin2017human}, the simulated demographic model did not apply demographic
    events in the past, so continental populations were simulated as isolated with low levels of migration
    for all time. The intended model is shown in Figure \ref{fig:ooa_stats}A.
    \textbf{(B-G)} We repeated the simulation experiment in~\cite{martin2017human} using the correct
    demographic model.
    While risk prediction in the African and East Asian population is reduced compared to the studied
    European population, the reduction in prediction accuracy is not as hopeless as originally reported.
    Notably, under a neutral polygenic architecture, we do not expect significant bias in inferred PRS
    in any of the populations (C, D).
    Correlations were computed over 100 simulation replicates.
    For direct comparison to the original study, see Figure~5 in~\citet{martin2017human}.
}
\label{fig:prs}
\end{center}
\end{figure}


\end{document}
